src/helpers/h.ts

```diff
@@ -0,0 +1,153 @@
+      const childNode = children[i];
+      if (childNode != null && isVElement(childNode)) {
+        addNS(childNode);
+      }
+    }
+  }
+}
+
+const { isArray } = Array;
+const { charCodeAt } = String.prototype;
+
+const sChar = 115, vChar = 118, gChar = 103, exclamationChar = 33;
+
+function primitive(s: any): s is (string | number) {
+  return typeof s === 'string' || typeof s === 'number';
+}
+
+const {ELEMENT_NODE, TEXT_NODE, COMMENT_NODE, DOCUMENT_FRAGMENT_NODE} = Node;
+
+export function vnodeElement(tag: string, data: VNodeData,
+                            children: VNodes, text?: string): VElement {
+  const { key } = data;
+  let sel = tag, elm;
+  return {
+    nt: ELEMENT_NODE,
+    tag,
+    sel,
+    data,
+    children,
+    text,
+    elm,
+    key
+  };
+}
+
+export function vnodeFragment(data: VNodeData, children: VNodes, text?: string): VFragment {
+  let sel, key, elm;
+  return {
+    nt: DOCUMENT_FRAGMENT_NODE,
+    sel,
+    data,
+    children,
+    text,
+    elm,
+    key,
+  };
+}
+
+export function vnodeText(text: string): VText {
+  let sel, data = {}, children, key, elm;
+  return {
+    nt: TEXT_NODE,
+    sel,
+    data,
+    children,
+    text,
+    elm,
+    key,
+  };
+}
+
+export function vnodeComment(text: string): VComment {
+  let sel = '!', data = {}, children, key, elm;
+  return {
+    nt: COMMENT_NODE,
+    sel,
+    data,
+    children,
+    text,
+    elm,
+    key,
+  };
+}
+
+function h(sel: string): VNode;
+function h(sel: string, data: VNodeData): VNode;
+function h(sel: string, text: string): VNode;
+function h(sel: string, children: VNodesSparse): VNode;
+function h(sel: string, data: VNodeData, text: string): VNode;
+function h(sel: string, data: VNodeData, children: VNodesSparse): VNode;
+function h(sel: any, b?: any, c?: any): VNode {
+  var data: VNodeData = {}, children: any, text: any, i: number;
+  if (c !== undefined) {
+    data = b;
+    if (isArray(c)) { children = c; }
+    else if (primitive(c)) { text = c; }
+    else if (c && c.sel) { children = [c]; }
+  } else if (b !== undefined) {
+    if (isArray(b)) { children = b; }
+    else if (primitive(b)) { text = b; }
+    else if (b && b.sel) { children = [b]; }
+    else { data = b; }
+  }
+  const firstCharCode = charCodeAt.call(sel, 0);
+  if (firstCharCode === exclamationChar) {
+    return vnodeComment(text == null ? '' : text as string);
+  } else if (sel === undefined) {
+    return vnodeText(text == null ? '' : text as string);
+  }
+  // normalize children
+  if (isArray(children)) {
+    for (i = 0; i < children.length; ++i) {
+      const ch = children[i];
+      if (ch == null) children[i] = null;
+      else if (primitive(ch)) children[i] = vnodeText(ch as string);
+    }
+  } else if (children != null) {
+    children = [children];
+  }
+  // normalize data
+  data = data || {};
+  if (sel === '') {
+    return vnodeFragment(data, children, text);
+  }
+  // parse selector
+  const hashIdx = sel.indexOf('#');
```

Comment:
I really like how you've moved the selector parsing into `h` while still making it possible to use id and static classes in the selector to guide the DOM diffing.